1-Parallelize-your-python-code_Futures_API.ipynb

```diff
@@ -0,0 +1,893 @@
+    "\n",
+    "However, we have a nice solution to be able to scrape more pages. When using a cluster each worker has its own public ip-address so it is like we are requesting from different machines. \n",
+    "\n",
+    "Let's create a coiled cluster and scrape a bigger number of pages:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a66cab38-f727-454d-b901-33f70b86a08b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Shutdown LocalCluster\n",
+    "client.shutdown()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9b81208e-1075-412a-ac48-e2193c984acb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import coiled"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6d91a1f5-b9e1-4553-b9c4-3da14127c59f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cluster = coiled.Cluster(n_workers=10, \n",
+    "                        package_sync=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4334ae4e-43a2-450e-a6fc-7f1b9c302f3f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "client =  Client(cluster)\n",
+    "client"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "401e1817-fe7a-4d3c-927e-44db2f5be74e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "pages_10_futures = client.map(get_page_html_links, range(1,11))\n",
+    "wait(pages_10_futures)\n",
+    "\n",
+    "posts_links_futures = client.map(get_post_links_per_page, pages_10_futures)\n",
+    "crawling = as_completed(posts_links_futures)\n",
+    "\n",
+    "dfs_data = []\n",
+    "for future in crawling:\n",
+    "    list_links = future.result() # list of links per page\n",
+    "    df_data = []\n",
+    "    for link in list_links:\n",
+    "        fut_data = client.submit(get_data, link) \n",
+    "        df_data.append(fut_data)\n",
+    "\n",
+    "    dfs_data.append(df_data)\n",
+    "_ = wait(dfs_data)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "01a64300-e971-4235-b9a9-b16c778c1069",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "len(dfs_data) #10 pages"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6fda0881-f0a1-47d0-b019-8f7c4270b5f7",
+   "metadata": {},
+   "source": [
+    "At this point, we have the data to build each page dataframe. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "67513dcf-0ee9-4981-a948-8f3697d0db0b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dfs_data[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1fa7fdfb-4a50-43ef-b662-76ccc892c141",
+   "metadata": {},
+   "source": [
+    "### Exercise \n",
+    "Using futures, transform `dfs_data` into a list of pandas.Dataframe futures."
```

Comment:
Why are we doing this?  The reader doesn't have any motivation at this point.  I think that in pedagogical situations like this it's very important to show them very clearly where they're going first, and then start to take steps in that direction.  I *think* that you're building up to a futures -> dask.dataframe transition (seems like a useful lesson) but that's not obvious at this point.